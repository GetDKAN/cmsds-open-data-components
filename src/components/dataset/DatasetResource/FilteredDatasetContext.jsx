import { createContext } from 'react'

export const FilteredDispatch = createContext(null)

export const defaultResourceState = {
  columnOrder: [],
  columns: [],
  count: 0,
  unfilteredCount: 0,
  unfilteredColumnCount: 0,
  currentPage: 0,
  density: 'density-3',
  excludedColumns: {},
  filters: [],
  loading: false,
  limit: 20,
  sort: [],
  values: [],
  downloadSize: {
    sizeLabel: 'KB',
    size: 0
  }
}
