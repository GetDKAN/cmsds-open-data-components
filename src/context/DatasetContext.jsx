import React, { useState, createContext } from 'react'
import PropTypes from 'prop-types'

export const DatasetContext = createContext()

export const DatasetProvider = ({ children }) => {
  const [datasetState, setDatasetState] = useState({
    data: {},
    error: null,
    isLoading: true
  })

  const resetDatasetState = () => {
    setDatasetState({
      data: {},
      error: null,
      isLoading: true
    })
  }

  const value = {
    ...datasetState,
    setDatasetState,
    resetDatasetState
  }

  return (
    <DatasetContext.Provider value={value}>{children}</DatasetContext.Provider>
  )
}

DatasetProvider.propTypes = {
  children: PropTypes.any
}
