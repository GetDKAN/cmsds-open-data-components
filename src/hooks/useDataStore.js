import { useState, useEffect, useRef } from 'react'
import axios from 'axios'
import qs from 'qs'

const fetchDataFromQuery = async (
  id,
  rootUrl,
  options,
  additionalParams
) => {
  const {
    keys,
    limit,
    offset,
    conditions,
    sort,
    groupings,
    prepareColumns,
    properties,
    unfilteredColumnCount,
    unfilteredCount,
    setValues,
    setUnfilteredColumnCount,
    setCount,
    setUnfilteredCount,
    setColumns,
    setLoading,
    setSchema
  } = options
  if (!id) {
    return false
  }
  if (typeof setLoading === 'function') {
    setLoading(true)
  }
  return await axios({
    method: 'GET',
    url: `${rootUrl}/datastore/query/${id}`,
    params: {
      keys,
      limit,
      offset,
      conditions,
      sorts: sort,
      properties,
      groupings,
      ...additionalParams
    },
    paramsSerializer: (params) => {
      return qs.stringify(params)
    }
  }).then((res) => {
    const { data } = res
    const propertyKeys =
      data.schema[id] && data.schema[id].fields
        ? Object.keys(data.schema[id].fields)
        : []

    setValues(data.results)
    setCount(data.count)

    if ([null, 0].includes(unfilteredCount)) {
      setUnfilteredCount(data.count)
    }

    if ([null, 0].includes(unfilteredColumnCount)) {
      const cols = prepareColumns ? prepareColumns(propertyKeys) : propertyKeys
      setUnfilteredColumnCount(cols.length)
    }

    if (propertyKeys.length) {
      setColumns(prepareColumns ? prepareColumns(propertyKeys) : propertyKeys)
    }
    setSchema(data.schema)

    if (typeof setLoading === 'function') {
      setLoading(false)
    }

    return data
  })
}

const useDatastore = (
  resourceId,
  rootAPIUrl,
  options,
  additionalParams = {}
) => {
  const keys = options.keys ? options.keys : true
  const { prepareColumns } = options
  const [manual, setManual] = useState(options.manual ? options.manual : false)
  const [requireConditions, setRequireConditions] = useState(
    options.requireConditions ? options.requireConditions : false
  )
  const [values, setValues] = useState([])
  const [id, setResource] = useState(resourceId)
  const [rootUrl, setRootUrl] = useState(rootAPIUrl)
  const [limit, setLimit] = useState(options.limit ? options.limit : 20)
  const [count, setCount] = useState(null)
  const [unfilteredCount, setUnfilteredCount] = useState(null)
  const [columns, setColumns] = useState([])
  const [unfilteredColumnCount, setUnfilteredColumnCount] = useState(null)
  const [offset, setOffset] = useState(options.offset ? options.offset : 0)
  const [loading, setLoading] = useState(false)
  const [conditions, setConditions] = useState(
    options.conditions ? options.conditions : undefined
  )
  const [sort, setSort] = useState(options.sort ? options.sort : undefined)
  const [groupings, setGroupings] = useState(
    options.groupings ? options.groupings : undefined
  )
  const [schema, setSchema] = useState({})
  // const [joins, setJoins] = useState()
  const [properties, setProperties] = useState(
    options.properties ? options.properties : undefined
  )
  const prevLimitRef = useRef()
  const prevOffsetRef = useRef()

  useEffect(() => {
    prevLimitRef.current = limit
    prevOffsetRef.current = offset
  })
  const prevLimit = prevLimitRef.current
  const prevOffset = prevOffsetRef.current

  function fetchData () {
    const newOffset = prevLimit === limit ? (prevOffset !== offset ? offset : 0) : 0

    setOffset(newOffset)
    fetchDataFromQuery(
      id,
      rootUrl,
      {
        keys,
        limit,
        offset: newOffset,
        conditions,
        sort,
        groupings,
        prepareColumns,
        properties,
        unfilteredColumnCount,
        unfilteredCount,
        setValues,
        setCount,
        setUnfilteredColumnCount,
        setUnfilteredCount,
        setColumns,
        setLoading,
        setSchema,
        setProperties
      },
      additionalParams
    )
  }

  useEffect(() => {
    if (!loading && !manual) {
      if (!requireConditions) {
        fetchData()
      } else if (requireConditions) {
        if (conditions && conditions.length) {
          fetchData()
        } else {
          setCount(null)
          setUnfilteredCount(null)
          setValues([])
        }
      }
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id, rootUrl, offset, conditions, sort, limit, requireConditions])

  return {
    loading,
    values,
    count,
    columns,
    limit,
    offset,
    schema,
    conditions,
    properties,
    unfilteredColumnCount,
    unfilteredCount,
    setProperties,
    setGroupings,
    setResource,
    setRootUrl,
    setLimit,
    setOffset,
    setConditions,
    setSort,
    setManual,
    setRequireConditions,
    fetchData
  }
}

export default useDatastore
