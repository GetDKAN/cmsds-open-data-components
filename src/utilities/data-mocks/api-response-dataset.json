{
    "fixture": true,
    "accessLevel": "public",
    "landingPage": "https://data.cms.gov/provider-data/dataset/test-data-abcd-0005",
    "bureauCode": [
        "009:38"
    ],
    "issued": "2020-01-20",
    "@type": "dcat:Dataset",
    "modified": "2020-01-20",
    "released": "2020-09-17",
    "archiveExclude": false,
    "keyword": [
        {
            "identifier": "8a0cf0a8-a3d0-5223-b1cc-82fbb3101cdb",
            "data": "hhc"
        },
        {
            "identifier": "9929edf6-e8af-5e74-b28e-6c43c0e0a534",
            "data": "general information"
        }
    ],
    "contactPoint": {
        "@type": "vcard:Contact",
        "fn": "Home Health Compare",
        "hasURL": "https://www.cms.gov/apps/contacts"
    },
    "publisher": {
        "identifier": "4c1de1c4-f083-5877-8968-7ddc04756c49",
        "data": {
            "@type": "org:Organization",
            "name": "Centers for Medicare & Medicaid Services (CMS)"
        }
    },
    "identifier": "test-data-abcd-0005",
    "description": "Home Health Care - Measure Date Range",
    "title": "Home Health Care - Measure Date Range",
    "programCode": [
        "009:000"
    ],
    "distribution": [
        {
            "identifier": "716256f0-6d17-56a6-9b5d-c35879011e9e",
            "data": {
                "@type": "dcat:Distribution",
                "downloadURL": "https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv",
                "mediaType": "text/csv",
                "%Ref:downloadURL": [
                    {
                        "identifier": "2850c4af459ff447d4e9bada0637ec54__1701355540__source",
                        "data": {
                            "filePath": "s3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0005.csv",
                            "identifier": "2850c4af459ff447d4e9bada0637ec54",
                            "mimeType": "text/csv",
                            "perspective": "source",
                            "version": "1701355540",
                            "checksum": null
                        }
                    },
                    {
                        "identifier": "2850c4af459ff447d4e9bada0637ec54__1701355540__local_url",
                        "data": {
                            "filePath": "https://h-o.st/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv",
                            "identifier": "2850c4af459ff447d4e9bada0637ec54",
                            "mimeType": "text/csv",
                            "perspective": "local_url",
                            "version": "1701355540",
                            "checksum": null
                        }
                    }
                ]
            },
            "format": "csv"
        }
    ],
    "theme": [
        {
            "identifier": "eda174d2-96df-5537-868f-f5ed625c3ede",
            "data": "Home health services"
        }
    ],
    "%modified": "2023-11-30T09:45:33-0500"
}
