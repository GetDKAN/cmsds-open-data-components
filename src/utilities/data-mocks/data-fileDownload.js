export const resourceData = {
  data: {
    identifier: '11745687-dc47-52ac-b9e4-66c50fe772ad',
    data: {
      '@type': 'dcat:Distribution',
      title: 'Test Data abcd 0016',
      downloadURL: 'https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/f69b073f217ea35ed14997178f7085c8_1677006237/Test_Data_abcd_0016.csv',
      mediaType: 'text/csv',
      '%Ref:downloadURL': [
        {
          identifier: 'f69b073f217ea35ed14997178f7085c8__1677006237__source',
          data: {
            filePath: 's3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0016.csv',
            identifier: 'f69b073f217ea35ed14997178f7085c8',
            mimeType: 'text/csv',
            perspective: 'source',
            version: '1677006237',
            checksum: null
          }
        },
        {
          identifier: 'f69b073f217ea35ed14997178f7085c8__1677006237__local_url',
          data: {
            filePath: 'http://h-o.st/provider-data/sites/default/files/resources/f69b073f217ea35ed14997178f7085c8_1677006237/Test_Data_abcd_0016.csv',
            identifier: 'f69b073f217ea35ed14997178f7085c8',
            mimeType: 'text/csv',
            perspective: 'local_url',
            version: '1677006237',
            checksum: null
          }
        }
      ]
    },
    format: 'csv'
  }
}

export const responseHeaders = {
  'content-length': 99999
}
