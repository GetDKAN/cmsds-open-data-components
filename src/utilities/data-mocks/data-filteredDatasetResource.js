export const filteredDatasetResource = {
  // identifier: '',
  // identifier: '07043497-a5ec-53f0-8a78-230dcd7f3e33',
  data: {
    '@type': 'dcat:Distribution',
    downloadURL: 'https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv',
    mediaType: 'text/csv',
    '%Ref:downloadURL': [
      {
        identifier: '2850c4af459ff447d4e9bada0637ec54__1701355540__source',
        data: {
          filePath: 's3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0005.csv',
          identifier: '2850c4af459ff447d4e9bada0637ec54',
          mimeType: 'text/csv',
          perspective: 'source',
          version: '1701355540',
          checksum: null
        }
      },
      {
        identifier: '2850c4af459ff447d4e9bada0637ec54__1701355540__local_url',
        data: {
          filePath: 'https://h-o.st/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv',
          identifier: '2850c4af459ff447d4e9bada0637ec54',
          mimeType: 'text/csv',
          perspective: 'local_url',
          version: '1701355540',
          checksum: null
        }
      }
    ]
  },
  format: 'csv'
}
