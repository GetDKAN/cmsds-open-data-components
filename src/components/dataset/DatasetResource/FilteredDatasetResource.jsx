import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { FilteredDispatch } from './FilteredDatasetContext'
import useDatastore from '../../../hooks/useDataStore'
import { getApiBaseUrl } from '../../../utilities/getApiBaseUrl'

import {
  useTable,
  usePagination,
  useFilters,
  useSortBy,
  useFlexLayout,
  useResizeColumns,
  useColumnOrder
} from 'react-table'

const FilteredDatasetResource = ({
  children,
  resource
}) => {
  const [filters, setFilters] = useState([{ column: '', condition: '', value: '' }]) // keep track of the applied filters
  const [filtersApplied, setFiltersApplied] = useState([])
  const [initCol, setInitCol] = useState([]) // we want to save a reference to filter against
  const [col, setCol] = useState([]) // columns we pass to table, post filter
  const [data, setData] = useState([]) // these are our cell values
  const [initOrder, setInitOrder] = useState([]) // initial order saved as a reference
  const [curOrder, setCurOrder] = useState([]) // current order that we pass to the data table
  const [visCol, setVisCol] = useState([]) // our visible column reference to prevent more datastore queries than need be
  const [activeDensity, setActiveDensity] = useState('Normal')
  const [downloadSize, setDownloadSize] = useState({
    sizeLabel: 'KB',
    size: 0
  })
  const prepareColumns = (columns, schema) => {
    return columns.map((column) => ({
      Header: schema && schema.fields[column].description ? schema.fields[column].description : column,
      accessor: column
    }))
  }
  const baseUrl = getApiBaseUrl()
  const filteredResource = useDatastore(
    '',
    baseUrl,
    {
      limit: 20,
      manual: true
    },
    {}
    // additionalParams
  )

  // listen to our filter resource changing, and flip the go switch once we have the info to query.
  useEffect(() => {
    if (resource.identifier) {
      filteredResource.setResource(resource.identifier)
      filteredResource.setManual(false)
    }
    if (!resource.identifier) {
      setInitOrder([])
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [resource])

  useEffect(() => {
    if (visCol.length) {
      const newCol = initCol.filter((item) => visCol.includes(item.accessor))
      setCol(newCol)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [visCol])

  // the issue is that this fires more than expected.
  useEffect(() => {
    if (Object.keys(filteredResource.schema).length > 0) {
      setData(filteredResource.values)
      // we only want to initialize columns if we don't have any info
      if (!initCol.length) {
        const newColumns = prepareColumns(filteredResource.columns, filteredResource.schema[Object.keys(filteredResource.schema)])
        setCol(newColumns) // init
        setInitCol(newColumns) // init
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filteredResource.loading])

  const filterTypes = React.useMemo(
    () => ({
      // Add a new fuzzyTextFilterFn filter type.
      // fuzzyText: fuzzyTextFilterFn,
      // Or, override the default text filter to use
      // "startWith"
      text: (rows, id, filterValue) => (
        rows.filter((row) => {
          const rowValue = row.values[id]
          return rowValue !== undefined
            ? String(rowValue)
              .toLowerCase()
              .startsWith(String(filterValue).toLowerCase())
            : true
        })
      )
    }),
    []
  )

  const defaultColumn = React.useMemo(
    () => ({
      // Let's set up our default Filter UI
      // Filter: DefaultColumnFilter,
      Filter: false,
      minWidth: 30,
      // width: 150,
      maxWidth: 400
    }),
    []
  )
  const filteredTable = useTable(
    {
      columns: col,
      data,
      initialState: { pageIndex: Number(filteredResource.offset / filteredResource.limit), pageSize: 20 },
      manualPagination: true,
      manualSortBy: true,
      manualFilters: true,
      pageCount: Math.ceil(Number(filteredResource.count) / filteredResource.limit),
      defaultColumn,
      filterTypes
      // NextPage and previous Page are defined methods in the Dataset explorer instead
    },
    useFilters,
    useFlexLayout,
    useResizeColumns,
    useColumnOrder,
    useSortBy,
    usePagination
  )

  // Initialize current and init orders
  useEffect(() => {
    if (filteredTable.allColumns.length && !initOrder.length) {
      setInitOrder(filteredTable.allColumns)
      setCurOrder(filteredTable.allColumns.map((d) => d.id))
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filteredTable.allColumns.length, initOrder.length])

  const resetVisibility = () => {
    const initialCol = initCol.map((item) => item.accessor)
    setVisCol(initialCol)
  }

  return (
    <FilteredDispatch.Provider
      value={{
        filteredResource,
        filteredTable,
        initOrder,
        curOrder,
        setCurOrder,
        setVisCol,
        resetVisibility,
        initCol,
        visCol,
        activeDensity,
        setActiveDensity,
        filters,
        setFilters,
        filtersApplied,
        setFiltersApplied,
        downloadSize,
        setDownloadSize
      }}
    >
      {children}
    </FilteredDispatch.Provider>
  )
}
FilteredDatasetResource.propTypes = {
  children: PropTypes.node,
  identifier: PropTypes.string,
  resource: PropTypes.object
}
FilteredDatasetResource.displayName = 'FilteredDatasetResource'
export default FilteredDatasetResource
