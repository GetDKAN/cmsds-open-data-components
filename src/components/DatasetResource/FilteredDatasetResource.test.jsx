import { useContext } from 'react';
import { render, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { DatasetContext } from '../../context/DatasetContext';
import FilteredDatasetResource from './FilteredDatasetResource';
import { filteredDatasetResource } from '../../utilities/data-mocks/data-filteredDatasetResource';
import { FilteredDispatch } from './FilteredDatasetContext';
import * as useDatastore from '../../hooks/useDataStore';

const mockDatasetResponse = require('../../utilities/data-mocks/api-response-dataset.json');

// Mock useDataStore hook value
const mockUseDataStoreDefaultValue = {
  loading: false,
  values: [
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '022',
      col4: '0',
      col5: '1745108',
      col6: 'M0023901',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '804',
      col4: '0',
      col5: '545293',
      col6: 'M0004994',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '804',
      col4: '0',
      col5: '545293',
      col6: 'M0012387',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '806',
      col4: '0',
      col5: '1251596',
      col6: 'M0004994',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '017',
      col4: '0',
      col5: '1482814',
      col6: 'M0004771',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '802',
      col4: '0',
      col5: '795085',
      col6: 'M0374010',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '802',
      col4: '0',
      col5: '848164',
      col6: 'M0023901',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '801',
      col4: '0',
      col5: '2200177',
      col6: 'M0006031',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '021',
      col4: '0',
      col5: '1653166',
      col6: 'M0023901',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '803',
      col4: '0',
      col5: '1720881',
      col6: 'M0004994',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '009',
      col4: '0',
      col5: '1745108',
      col6: 'M0023901',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '022',
      col4: '0',
      col5: '795085',
      col6: 'M0005335',
    },
    {
      col1: '00022225',
      col2: 'S5660',
      col3: '804',
      col4: '0',
      col5: '545293',
      col6: 'M0019437',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '004',
      col4: '0',
      col5: '795085',
      col6: 'M0001750',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '801',
      col4: '0',
      col5: '1653166',
      col6: 'M0004771',
    },
    {
      col1: '00022509',
      col2: 'H5050',
      col3: '019',
      col4: '0',
      col5: '795085',
      col6: 'M0374010',
    },
  ],
  count: 16,
  columns: ['col1', 'col2', 'col3', 'col4', 'col5', 'col6'],
  limit: 20,
  offset: 0,
  schema: {
    '7d48577d-2944-56f7-bb48-2f8272e87007': {
      fields: {
        col1: {
          type: 'text',
          mysql_type: 'text',
        },
        col2: {
          type: 'text',
          mysql_type: 'text',
        },
        col3: {
          type: 'text',
          mysql_type: 'text',
        },
        col4: {
          type: 'text',
          mysql_type: 'text',
        },
        col5: {
          type: 'int',
          length: 11,
          mysql_type: 'int',
          description: 'Column 5',
        },
        col6: {
          type: 'text',
          mysql_type: 'text',
        },
      },
    },
  },
  conditions: undefined,
  properties: undefined,
  setProperties: vi.fn(),
  setGroupings: vi.fn(),
  setResource: vi.fn(),
  setRootUrl: vi.fn(),
  setLimit: vi.fn(),
  setOffset: vi.fn(),
  setConditions: vi.fn(),
  setSort: vi.fn(),
  setManual: vi.fn(),
  setRequireConditions: vi.fn(),
  fetchData: vi.fn(),
};

const datasetContextProviderValue = {
  data: mockDatasetResponse,
  error: null,
  isLoading: true,
  setDatasetState: vi.fn(),
  resetDatasetState: vi.fn(),
};

let datasetState;
const renderComponent = ({
  contextValue = datasetContextProviderValue,
  resource = filteredDatasetResource,
  getStateValue = (data) => {
    datasetState = data;
  },
}) => {
  return render(
    <DatasetContext.Provider value={contextValue}>
      <FilteredDatasetResource resource={resource}>
        <FilteredDatasetResourceTestingComponent getStateValue={getStateValue} />
      </FilteredDatasetResource>
    </DatasetContext.Provider>
  );
};

const FilteredDatasetResourceTestingComponent = ({ getStateValue }) => {
  getStateValue(useContext(FilteredDispatch));
};

describe('FilteredDatasetResource component', () => {
  beforeEach(() => {
    datasetState = null;

    vi.spyOn(useDatastore, 'default').mockImplementation(() => mockUseDataStoreDefaultValue);
  });

  it('Matches snapshot', () => {
    const renderedFilteredDatasetResource = renderer
      .create(
        <DatasetContext.Provider value={datasetContextProviderValue}>
          <FilteredDatasetResource resource={filteredDatasetResource}>
            Dataset content goes here
          </FilteredDatasetResource>
        </DatasetContext.Provider>
      )
      .toJSON();

    expect(renderedFilteredDatasetResource).toMatchSnapshot();
  });

  it('Dataset state columns are correct', () => {
    renderComponent({});

    expect(datasetState.filteredResource.columns).toEqual([
      'col1',
      'col2',
      'col3',
      'col4',
      'col5',
      'col6',
    ]);
  });

  it('Dataset state columns are empty if useDatastore returns none', () => {
    vi.spyOn(useDatastore, 'default').mockImplementation(() => ({
      loading: false,
      values: [],
      count: 0,
      columns: [],
      limit: 20,
      offset: 0,
      schema: {},
      conditions: undefined,
      properties: undefined,
      setProperties: vi.fn(),
      setGroupings: vi.fn(),
      setResource: vi.fn(),
      setRootUrl: vi.fn(),
      setLimit: vi.fn(),
      setOffset: vi.fn(),
      setConditions: vi.fn(),
      setSort: vi.fn(),
      setManual: vi.fn(),
      setRequireConditions: vi.fn(),
      fetchData: vi.fn(),
    }));

    const mockResource = Object.assign({}, filteredDatasetResource);
    delete mockResource.identifier;

    renderComponent({
      resource: mockResource,
    });

    expect(datasetState.filteredResource.columns).toEqual([]);
  });

  it('Dataset columns are properly updated when filtered', async () => {
    renderComponent({});

    // Filter the columns
    await act(async () => {
      datasetState.filteredResource.loading = true; // This is just for full test coverage
      datasetState.setVisCol(['col1', 'col2']);
    });

    const filteredTableColumns = [];
    datasetState.filteredTable.columns.forEach((column) => {
      filteredTableColumns.push(column.id);
    });

    expect(datasetState.visCol).toEqual(['col1', 'col2']);
    expect(filteredTableColumns).toEqual(['col1', 'col2']);

    // Reset the columns
    await act(async () => {
      datasetState.filteredResource.loading = false; // This is just for full test coverage
      datasetState.resetVisibility();
    });

    const resetTableColumns = [];
    datasetState.filteredTable.columns.forEach((column) => {
      resetTableColumns.push(column.id);
    });

    expect(datasetState.visCol).toEqual(['col1', 'col2', 'col3', 'col4', 'col5', 'col6']);
    expect(resetTableColumns).toEqual(['col1', 'col2', 'col3', 'col4', 'col5', 'col6']);
  });

  it('Data table rows are properly filtered when triggered', () => {
    renderComponent({});

    // This is just to get full test coverage
    const modifiedRows = { ...datasetState.filteredTable.rows };
    modifiedRows[1].values['col6'] = undefined;

    // Filter table rows at column 6 with 'M0023901' filter - uses 'starts with'
    const filteredRows = datasetState.filteredTable.filterTypes.text(
      datasetState.filteredTable.rows,
      'col6',
      'M0023901'
    );
    const expectedRowIds = ['0', '1', '6', '8', '10'];
    const actualRowIds = [];

    filteredRows.forEach((row) => {
      actualRowIds.push(row.id);
    });

    expect(filteredRows.length).toEqual(5);
    expect(actualRowIds).toEqual(expectedRowIds);
  });
});
