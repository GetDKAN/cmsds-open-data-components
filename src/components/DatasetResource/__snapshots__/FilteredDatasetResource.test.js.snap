// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredDatasetResource component Matches snapshot 1`] = `"Dataset content goes here"`;
